ci/webhook-receiver.py

```diff
@@ -196,13 +204,30 @@ def init_ci_pipeline():
     out, err = proc2.communicate()
 
     if (out):
-        print("Can't run CI, another CI run in progress")
-        log_access_granted(request_ctx, "CI busy, posting busy msg")
-        os.system("./ci_busy.sh config {} \"{}\" \"{}\" \"Another CI run in progress, please try again in 15 minutes\""
-                  .format(request_ctx['id'], request_ctx['repo'], request_ctx['body']))
+        print("Can't run CI, another CI run in progress, placing in the queue")
+        log_access_granted(request_ctx, "CI busy, placing request in queue")
+        duplicate_req = False
+        busy_msg = "Duplicate request already waiting, ignoring message"
+        with queue_lock:
+            # make sure we don't bypass thread safety
+            for req in ci_list:
+                if req['id'] == request_ctx['id'] and req['repo'] == request_ctx['repo']:
+                    # make sure this is the same PR
+                    duplicate_req = True
+                    break
+            if not duplicate_req:
+                # don't have this request yet, put in queue
+                ci_list.append(request_ctx)
+                busy_msg = "Another CI run in progress, adding request to the end of the list"
+        # ending frees the lock
+        os.system("./ci_busy.sh config {} \"{}\" \"{}\" \"{}\""
+                  .format(request_ctx['id'], request_ctx['repo'], request_ctx['body'], busy_msg))
     else:
-        log_access_granted(request_ctx, "Running CI")
-        os.system("./manager.sh config {} \"{}\" \"{}\"".format(request_ctx['id'], request_ctx['repo'], request_ctx['body']))
+        # no manager running yet, run with current request
+        run_manager(request_ctx)
+        while len(ci_list) > 0:
+            # new requests were made since our last call, run them in order
+            run_manager(ci_list.pop(0))
```

Comment:
Even though I think this is safe (pop is atomic right?) the design isn't very elegant. Would it be possible for us to have a different thread poll the queue and run things when items appear there?